src/Tempest/Http/src/Routing/Construction/RouteTreeNode.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Http\Routing\Construction;
+
+use Tempest\Http\Route;
+
+/**
+ * @internal
+ */
+final class RouteTreeNode
+{
+    /** @var array<string, RouteTreeNode> */
+    private array $staticPaths = [];
+
+    /** @var array<string, RouteTreeNode> */
+    private array $dynamicPaths = [];
+
+    private ?MarkedRoute $leaf = null;
+
+    private function __construct(
+        public readonly RouteTreeNodeType $type,
+        public readonly ?string $segment = null
+    ) {
+    }
+
+    public static function createRootRoute(): self
+    {
+        return new self(RouteTreeNodeType::Root);
+    }
+
+    public static function createParameterRoute(string $regex): self
+    {
+        return new self(RouteTreeNodeType::Parameter, $regex);
+    }
+
+    public static function createStaticRoute(string $name): self
+    {
+        return new self(RouteTreeNodeType::Static, $name);
+    }
+
+    public function addPath(array $pathSegments, MarkedRoute $markedRoute): void
```

Comment:
Adds a new route to the Routing tree. it uses recursion to do so. so with each step we take the first element of `$pathSegments` , look up the appropriate node (or create it) and recurse into it's addPath with the remaining `pathSegments`.

If `$pathSegments` is empty, we mark said node as a valid "leaf" node and attach the `MarkedRoute` to it

Reply:
If I followed my gut, my assumption would be that we should pass in the `string $path` instead of `array $pathSegments`, and split it internally. But there might be good reasons to use the part segments that I'm overlooking.

Reply:
Another sidenote: in `Route` it's called `routeParts`, while over here it's called `pathSegments`, I think it'd be best to make the name consistent — or maybe I'm confused and mixing two different concepts 😁

Reply:
I will stick to `pathSegments` as it is more describtive. I renamed the Route method to `split` as suggested and put `segments` into the explaining doc block.

> my assumption would be that we should pass in the string $path instead of array $pathSegments, and split it internally

It would be possible, but realllly wasteful. The RouteTreeNode is already the internal of the tree, so normally people will not see / use this. The reason why I use an array is because use recursion for each element in the `$pathSegments` array. You can see on me using `array_shift` to get the first element (and mutate the existing array) and passing the remaining array into the addPath of the next node. 

So when I start with the segments: `['test', '{id}', 'edit']` 
- I first find or create the node for `test`, and call `addPath` with the remaining segments: `['{id}', 'edit'], 
- then I find or create the node for `{id}` in the `test` node and call `addPath` with the remaining segments:  `['edit']`, 
- then I find or create edit `edit` node in the `{id}` node and call `addPath` with the remaining segments `[]`
- Last I add the passed along `$markedRoute` into the `$leaf`

Reply:
Fair enough, thanks for the clarification!